### PR TITLE
[InstCombine] (uitofp bool X) * Y --> X ? Y : 0

### PR DESCRIPTION
**Pull request by AlexMaclean**
Fold `mul (uitofp i1 X), Y` to `select i1 X, Y, 0.0` when the `mul` is `nnan` and `nsz`

Proof: https://alive2.llvm.org/ce/z/_stiPm

### CODE REVIEW

**Review by dtcxzyw — APPROVED** [review]
LGTM.

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp [hunk 1]
```diff
@@ -879,6 +879,16 @@ Instruction *InstCombinerImpl::visitFMul(BinaryOperator &I) {
...
+  if (I.hasNoNaNs() && I.hasNoSignedZeros()) {
+    // (uitofp bool X) * Y --> X ? Y : 0
+    // Y * (uitofp bool X) --> X ? Y : 0
+    // Note INF * 0 is NaN.
+    if (match(Op0, m_UIToFP(m_Value(X))) && X->getType()->isIntOrIntVectorTy(1))
+      return SelectInst::Create(X, Op1, ConstantFP::get(I.getType(), 0.0));
+    if (match(Op1, m_UIToFP(m_Value(X))) && X->getType()->isIntOrIntVectorTy(1))
+      return SelectInst::Create(X, Op0, ConstantFP::get(I.getType(), 0.0));
+  }
+
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:891]: You don't need nsz if you also introduce a copysign:

https://alive2.llvm.org/ce/z/Be9G_s

File: llvm/test/Transforms/InstCombine/fmul-bool.ll [hunk 2]
```diff
@@ -0,0 +1,35 @@
+; RUN: opt < %s -passes=instcombine -S | FileCheck %s
+
+; X * Y (when Y is a boolean) --> Y ? X : 0
+
+define float @fmul_bool(float %x, i1 %y) {
+; CHECK-LABEL: @fmul_bool(
+; CHECK-NEXT:    [[M:%.*]] = select i1 [[Y:%.*]], float [[X:%.*]], float 0.000000e+00
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fmul-bool.ll:7]: You can preserve the flags here

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp (hunk 1 above)
Comment (arsenm) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:891]: Not obvious to me if the form with copysign is better or not, it's 2 instructions -> 2 instructions

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:891]: With copysign we need to check that `uitofp` is one-use.

File: llvm/test/Transforms/InstCombine/fmul-bool.ll (hunk 2 above)
Comment (AlexMaclean) [reply llvm/test/Transforms/InstCombine/fmul-bool.ll]: Good point, fixed

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp (hunk 1 above)
Comment (AlexMaclean) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:899]: Is it alright to just start with this case?

**Review by arsenm — APPROVED** [review]

File: llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp (hunk 1 above)
Comment (arsenm) [reply llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:899]: Yes

**Pull request merged**